lib/tpm2_util.c

```diff
@@ -1100,11 +1102,142 @@ TPMI_ALG_HASH *param_hash_algorithm, tpm2_session **session) {
                 tpm2_session_get_handle(session[session_idx]);
         }
     }
+
+    return tool_rc_success;
+}
+
+#define SESSION_PREFIX "session:"
+#define SESSION_PREFIX_LEN sizeof(SESSION_PREFIX) - 1
+static bool adjust_sessions_auths_or_paths(const char **sessions_auths_or_paths){
+
+    size_t session_idx = 0;
+    size_t input_session_count = 0;
+    for (session_idx = 0; session_idx < MAX_SESSION_CNT; session_idx++) {
+        /*
+         * Sessions specified as aux sessions.
+         */
+        if (sessions_auths_or_paths[session_idx]) {
+            input_session_count++;
+        } else {
+            continue;
+        }
+
+        /*
+         * Sessions specified through the auth interface.
+         */
+        if(!strncmp(sessions_auths_or_paths[session_idx], SESSION_PREFIX,
+        SESSION_PREFIX_LEN)) {
+            sessions_auths_or_paths[session_idx] += SESSION_PREFIX_LEN;
+        }
+    }
+
+    if (!input_session_count) {
+        return false;
+    }
+
+    return true;
+}
+
+static TPMI_ALG_HASH calc_cphash_alg_from_session_types(ESYS_CONTEXT *ectx,
+const char **adjusted_session_paths) {
+
+    TPMI_ALG_HASH rethash = TPM2_ALG_SHA256;
+    size_t session_idx = 0;
+    for (session_idx = 0; session_idx < MAX_SESSION_CNT; session_idx++) {
+        if (!adjusted_session_paths[session_idx]) {
+            continue;
+        }
+
+        tpm2_session *session = NULL;
+        tool_rc rc = tpm2_session_restore(ectx,
+        adjusted_session_paths[session_idx], false, &session);
+        if (rc != tool_rc_success) {
+            LOG_WARN("Could not restore session #%s",
+            adjusted_session_paths[session_idx]);
+            tpm2_session_close(&session);
+            continue;
+        }
+
+        TPM2_SE session_type = tpm2_session_get_type(session);
+        if (session_type != TPM2_SE_HMAC && session_type != TPM2_SE_POLICY) {
+            tpm2_session_close(&session);
+            continue;
+        }
+
+        rethash = tpm2_session_get_authhash(session);
+
+        TPMA_SESSION attrs = 0;
+        ESYS_TR session_handle = tpm2_session_get_handle(session);
+        tool_rc tmp_rc = tpm2_sess_get_attributes(ectx, session_handle, &attrs);
+        UNUSED(tmp_rc);
```

Comment:
this is okay to Fail?

Reply:
Correct. If that happens we return the default SHA256.